src/session/session_api.c

```diff
@@ -674,6 +671,32 @@ __session_alter(WT_SESSION *wt_session, const char *uri, const char *config)
     API_END_RET_NOTFOUND_MAP(session, ret);
 }
 
+/*
+ * __session_alter --
+ *     Alter a table setting.
+ */
+static int
+__session_alter(WT_SESSION *wt_session, const char *uri, const char *config)
+{
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+
+    session = (WT_SESSION_IMPL *)wt_session;
+
+    /*
+     * Alter table can return EBUSY error when the table is modified in parallel by eviction. Retry
+     * the command after performing a system wide checkpoint. Don't retry it in a loop as it may
+     * lead to wait forever.
+     */
+    ret = __session_alter_internal(session, uri, config);
+    if (ret == EBUSY) {
+        WT_RET(session->iface.checkpoint(&session->iface, NULL));
+        ret = __session_alter_internal(session, uri, config);
+    }
+
+    return (ret);
+}
+
```

Comment:
Because we've given up the checkpoint lock between the alter attempt and checkpoint, there could already be a checkpoint running. Compaction has to deal with this, see `__compact_checkpoint`, but maybe something like this?

Compaction also has a timeout, that is, it won't wait longer than a configured amount of time, and that might be reasonable here as well so we don't stall  a thread for too long, we could pass in a max-timeout argument to this function?

If you agree, I suspect compaction could use this worker function too by calculating a maximum timeout to pass into this function.

I'm not sure this is really necessary, but it wouldn't cost us a lot of effort.
```
 * __session_blocking_checkpoint_ --
 *     Perform a checkpoint to resolve an EBUSY error.
 */
static int
__session_blocking_checkpoint_(WT_SESSION_IMPL *session, u_int seconds)
{   
    WT_DECL_RET;
    WT_TXN_GLOBAL *txn_global;
    uint64_t txn_gen;
    const char *checkpoint_cfg[] = {WT_CONFIG_BASE(session, WT_SESSION_checkpoint), NULL}; 

    if ((ret = __wt_txn_checkpoint(session, checkpoint_cfg, false)) == 0)
        return (0);
    WT_RET_BUSY_OK(ret);
 
    /*
     * If there's a checkpoint running, wait for it to complete. If there's no checkpoint running 
     * or the checkpoint generation number changes, the checkpoint blocking us has completed.  
     */
    txn_global = &S2C(session)->txn_global;
    for (txn_gen = __wt_gen(session, WT_GEN_CHECKPOINT);; __wt_sleep(2, 0), seconds -= 2) {
        /*
         * This loop only checks objects that are declared volatile, therefore no barriers are
         * needed.
         */
        if (!txn_global->checkpoint_running || txn_gen != __wt_gen(session, WT_GEN_CHECKPOINT))
            break;
 
        /* If there's a timeout, give up. */
        if (seconds == 0)
            continue;
        if (seconds <= 2)
            return (EBUSY);
    }
 
    return (0);
}
```

Reply:
Thanks. I will check and modify as per the suggestion.